README.md
# 𝓛𝓮-inventory

![Logo](docs/le-inventory.png)

![Tags](https://badgen.net/badge/icon/%23LeInventory%20%23py/14406F1?icon=https://icons.getbootstrap.com/assets/icons/bookmarks-fill.svg&label&labelColor=FFF)
![ProjectName](https://badgen.net/badge/Project%20Name/Le-Inventory/0058E7?labelColor=000) ![Version](https://badgen.net/badge/Version/01.01/cyan?labelColor=000) ![Authors](https://badgen.net/badge/Authors/CPR%20Ninis%20Alescs/60C?labelColor=000)

---

## Table of Contents

- [Description](#description)
- [Requirements](#requirements)
- [Directories structure](#directories-structure)
- [Execution examples](#execution-examples)

---

## Description

Le-Inventory is an inventory system, custom-made for the HW team.
It's used to optimize hardwaristi's job.

---

## Requirements

-  ![Language](https://badgen.net/badge/Python/v3.9+/FFD343?labelColor=3776AB&icon=pypi) <sup>([Download](https://www.python.org/downloads/)) <small>[it has to be in the PATH]</small></sup>
---

## Directories structure
	- data

	- docs
		- CHANGELOG.txt
	- requirements
		- requirements.txt
	- src
		- main.py
		- structures.py

	- .gitignore
	- LICENSE
	- README.md
---

## Execution examples

![CLI](https://badgen.net/badge/icon/CLI/B67DFF?icon=terminal&label&labelColor=000)
```bash
# syntax
path/to/le-inventory>py src/main.py [-h] [-i FILEIN] [-o FILEOUT] [-t {Resistor,Capacitor,Diode,Project} [-v]

# example

# 0 - help menu
python3 src/main.py -h
usage: le-inventory/src/main.py [-h] [-i FILEIN] [-o FILEOUT] [-t {Resistor,Capacitor,Diode,Project}] [-v]

A program to manage the stock of electronic components

options:
  -h, --help            show this help message and exit
  -i FILEIN, --filein FILEIN
                        input file
  -o FILEOUT, --fileout FILEOUT
                        output file
  -t {Resistor,Capacitor,Diode,Project}, --type {Resistor,Capacitor,Diode,Project}
                        type of the objects to add
  -v, --verbose         increase output verbosity

# 1 - non-verbose
path/to/le-inventory>python3 src/main.py -i input -o - Capacitors

# 2 - verbose
path/to/le-inventory>python3 src/main.py -i input -o resistors -t Resistor -v

Starting program...
Initializing variables...
Input file: input.csv
Output file: resistors.csv
Objects type: Resistor

Reading csv file data/input.csv...
Creating objects...
Adding object(s) to the dataframe...
Adding object(s) to the dataframe...
Done adding objects to the dataframe!

Serializing data to csv file: data/resistors.csv...
Execution was successful!
```

---

Made with ♡ by [![CPR](https://badgen.net/badge/icon/CPR/B67DFF?icon=github&label&labelColor=000)](https://github.com/chiarasabaini) [![Ninis](https://badgen.net/badge/icon/Ninis/B67DFF?icon=github&label&labelColor=000)](https://github.com/thomasnonis) [![Alescs](https://badgen.net/badge/icon/Alescs/B67DFF?icon=github&label&labelColor=000)](https://github.com/Alescs)


docs/CHANGELOG.txt
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [01.01] - 2020-3-24
 - first version

src/structures.py
__authors__ = "CPR, Ninis, Alescs"
__version__ = "01.01"
__date__ = "2023-03-16"

from dataclasses import dataclass

@dataclass
class Component():
    """
    Class used to represent a component

    ...

    Attributes
    ----------
    id : int
        component id
    PN : str
        part number
    location : str
        location in the warehouse
    package : str
        package type
    stock : int
        number of components in stock
    reserved : int
        number of components reserved for projects
    price : int
        price of one component
    
    Methods
    -------
    get_available()
        Returns the number of available components
    """    
    id : int
    PN : str
    location : str
    package : str
    stock : int
    reserved : int
    price : int


    def get_available(self):
        """
        Returns the number of available components

        Returns
        -------
        int
            number of available components
        """  
        return self.stock - self.reserved
    


@dataclass
class Resistor(Component):
    """
    Class used to represent a resistor

    ...

    Attributes
    ----------
    value : float
        resistance
    tolerance : float
        tolerance
    power : float
        power
    """
    value : float
    tolerance : float
    power : float
    

@dataclass
class Capacitor(Component):
    """
    Class used to represent a capacitor

    ...

    Attributes
    ----------
    value : float
        capacitance
    tolerance : float
        tolerance
    dielectric : str
        dielectric type
    maxVoltage : int
        maximum voltage
    """
    value : float
    tolerance : float
    dielectric : str
    maxVoltage : int


@dataclass
class Diode(Component):
    """
    Class used to represent a diode

    ...

    Attributes
    ----------
    type : str
        diode type
    value : float
        forward voltage
    power : float
        power
    forwardVoltage : float
        forward voltage
    breakdownVoltage : float
        breakdown voltage
    """
    type : str
    value : float
    power : float
    forwardVoltage : float
    breakdownVoltage : float


@dataclass
class Project():
    """
    Class used to represent a project

    ...

    Attributes
    ----------
    name : str
        project name
    version : str
        project version
    date : str
        project date
    projectPath : str
        path to the project
    resistors : list[(Resistor, int)]
        list of tuples, composed of the id of the resistors used in the project and the quantity
    capacitors : list[(Capacitor, int)]
        list of tuples, composed of the id of the capacitors used in the project and the quantity
    diodes : list[(Diode, int)]
        list of tuples, composed of the id of the diodes used in the project and the quantity
    
        
    Methods
    -------
    get_id()
        Returns the project id
    """
    name : str
    version : str
    date : str
    projectPath : str
    resistors : list[(Resistor, int)]
    capacitors : list[(Capacitor, int)]
    diodes : list[(Diode, int)]


    def get_id(self):
        """
        Returns the project id

        Returns
        -------
        str
            project id
        """        
        return f"{self.name}_{self.date}_{self.version.replace('.', '-')}"


src/main.py
__authors__ = "CPR, Ninis, Alescs"
__version__ = "01.01"
__date__ = "2023-03-16"


import argparse
import pandas as pd
import dataclasses as dc
from structures import *

VERBOSE = False

def add_to_df(df, objs):
    """"
    Adds a component or a list of components to a dataframe, then returns the dataframe

    Parameters
    ----------
    df : pandas.DataFrame
        dataframe to add the component(s) to
    objs : structures.Component or list of structures.Component
        component(s) to add to the dataframe

    Returns
    -------
    pandas.DataFrame
        dataframe with the added component(s)
    """

    if VERBOSE:
        print("Adding object(s) to the dataframe...")

    if(type(objs) == list):
        for o in objs:
            dict = dc.asdict(o)
            df.loc[len(df), dict.keys()] = dict
    else:
        dict = dc.asdict(objs)
        df.loc[len(df), dict.keys()] = dict

    return df


def add_from_csv(path, type):
    """"
    Creates objects from a csv file, adds them to a dataframe, and returns the dataframe

    Parameters
    ----------
    path : str
        path to the csv file
    type : str
        type of the objects to create

    Returns
    -------
    pandas.DataFrame
        dataframe from the csv file
    """
    if VERBOSE:
        print(f"Reading csv file {path}...")
    
    match type:
        case "Resistor":
            constructor = Resistor
        case "Capacitor":
            constructor = Capacitor
        case "Diode":
            constructor = Diode
        case "Project":
            constructor = Project

    if VERBOSE:
        print("Creating objects...")

    df = pd.DataFrame()
    with open(path, "r") as f:
        header = True
        for line in f:
            if header:
                header = False
                continue

            line = line.strip("\n").split(",")
            obj = constructor(line[0], line[1], line[2], line[3], line[4], line[5], line[6], line[7], line[8], line[9])
            add_to_df(df, obj)
    
    if VERBOSE:
        print("Done adding objects to the dataframe!\n")
    
    return df


def get_args():
    """
    Parses the arguments passed to the program

    Returns
    -------
    argparse.Namespace
        arguments passed to the program
    """
    
    parser = argparse.ArgumentParser(prog="le-inventory/src/main.py",
                                     description="A program to manage the stock of electronic components")

    parser.add_argument('-i', '--filein', help="input file")
    parser.add_argument('-o', '--fileout', help="output file")
    parser.add_argument('-t', '--type', help="type of the objects to add", choices=["Resistor", "Capacitor", "Diode", "Project"])
    parser.add_argument('-v', '--verbose', help="increase output verbosity", action="store_true", default=False)
    
    args = parser.parse_args()

    return args


def __init__():
    global VERBOSE
    args = get_args()

    fin = args.filein
    fout = args.fileout
    obj_type = args.type
    VERBOSE = args.verbose

    if VERBOSE:
        print(f"Starting program...\nInitializing variables...")
        print(f"Input file: {args.filein}.csv\nOutput file: {args.fileout}.csv\nObjects type: {args.type}\n")

    resistors_df = add_from_csv(f"data/{fin}.csv", f"{obj_type}")

    if VERBOSE:
        print(f"Serializing data to csv file: data/{fout}.csv...")
    
    if obj_type == "Project":
        path_out = f"data/{fout}.csv" # TODO: handle project type [filename must be the project's id, retrieved using Project.get_id()]
    else:
        path_out = f"data/{fout}.csv"
    resistors_df.to_csv(path_out, mode='a', index=False, header=False)
    
    if VERBOSE:
        print("Execution was successful!")

if __name__ == '__main__':  

    __init__()













